### PR TITLE
Convert MIDebugEngine to use SDK csproj

### PR DESCRIPTION
**Pull request by WardenGnaw**
This PR converts all the C# projects to use the new styled [sdk projects](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/overview).

Removed prebuild.csproj. The IL used for each specific project is added and is generated by the target which will run before resolving References.

| Status | Task |
| --- | :-- |
| ✔️ | Analyzers |
| ✔️ | Validate MicroBuild.Core |
| ✔️ | Fix Build Pipelines | 
| ✔️| Fix ` The literal or constant value <...> should be passed as the  'expected' argument in the call to 'Assert.Equal(expected, actual)' ` in tests |
| ✔️ | Fix MIDebugPackage.pkgdef Generation | 
| ✔️|  VSIX generation for SSHDebugPS |

TODO List:
| Status | Task |
| --- | :-- |
| 😖 | Fix MIDebugPackagePackage Threading Issues |
| :x: |   Remove `<PropertyGroup Condition = $(Configuration) \| $(Platform) == Desktop.Debug\|AnyCPU>` |

### CODE REVIEW

**Comment by gregg-miskelly** [comment]
>     <IntermediateGeneratedOutputPath>$(IntermediateOutputPath)Generated\</IntermediateGeneratedOutputPath>

Is this still used?

---
Refers to: build/all_projects.settings.targets:23 in ed816ac. [](commit_id = ed816ac959e1ada67b0834052c2248183fd0c79e, deletion_comment = False)

File: src/AndroidDebugLauncher/AndroidDebugLauncher.csproj [hunk 1]
```diff
@@ -48,7 +46,14 @@
     <ErrorReport>prompt</ErrorReport>
     <CodeAnalysisRuleSet>MinimumRecommendedRules.ruleset</CodeAnalysisRuleSet>
   </PropertyGroup>
-  <ItemGroup>
+
+  <ItemGroup Label="Assembly to Generate">
+    <GenerateAssembly Include="$(ILDir)libadb.il">
+      <Visible>false</Visible>
```
Comment (gregg-miskelly) [inline src/AndroidDebugLauncher/AndroidDebugLauncher.csproj:52]: ><Visible>false</Visible> [](start = 6, length = 24)

What winds up consuming `Visible`?

**Comment by gregg-miskelly** [comment]
Your new .pkgdef file somehow wound up being binary, even though MIEngine has the right .gitattributes file.

File: src/MIDebugPackage/MIDebugPackage.csproj [hunk 2]
```diff
@@ -174,10 +151,6 @@
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
       <IncludeInVSIX>true</IncludeInVSIX>
     </Content>
-    <None Include="packages.config">
-      <SubType>Designer</SubType>
-    </None>
-    <None Include="packages.props" />
```
Comment (gregg-miskelly) [inline src/MIDebugPackage/MIDebugPackage.csproj:180]: >packages.props [](start = 19, length = 14)

I think you forgot to delete packages.props

**Comment by gregg-miskelly** [comment]
>   </PropertyGroup>

Doesn't need to be done on this checkin, but should we delete all this crap?

---
Refers to: src/MIDebugPackage/MIDebugPackage.csproj:23 in ed816ac. [](commit_id = ed816ac959e1ada67b0834052c2248183fd0c79e, deletion_comment = False)

**Comment by gregg-miskelly** [comment]
Otherwise LGTM

**Comment by WardenGnaw** [comment]
@gregg-miskelly Is it normal to require `dotnet restore` before building the solution? The building experience is a bit weird in VS since you have to do the restore first before building the solution. 

There also seems to be an issue with `MakePIAPortable` and `WindowsDebugLauncher` not restoring when you run `dotnet restore src\MIDebugEngine.sln`

File: src/AndroidDebugLauncher/AndroidDebugLauncher.csproj (hunk 1 above)
Comment (WardenGnaw) [reply src/AndroidDebugLauncher/AndroidDebugLauncher.csproj:52]: Thought this was be for solution explorer. I will double check.

**Comment by WardenGnaw** [comment]
>Doesn't need to be done on this checkin, but should we delete all this crap?

Are you referencing the 
```
<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <DefineConstants>$(DefineConstants);DEBUG;TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
    <PlatformTarget>AnyCPU</PlatformTarget>
  </PropertyGroup>
```

**Comment by gregg-miskelly** [comment]
> > Doesn't need to be done on this checkin, but should we delete all this crap?
> 
> Are you referencing the
> 
> ```
> <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
>     <DebugSymbols>true</DebugSymbols>
>     <DebugType>full</DebugType>
>     <Optimize>false</Optimize>
>     <DefineConstants>$(DefineConstants);DEBUG;TRACE</DefineConstants>
>     <ErrorReport>prompt</ErrorReport>
>     <WarningLevel>4</WarningLevel>
>     <PlatformTarget>AnyCPU</PlatformTarget>
>   </PropertyGroup>
> ```

Yes

**Comment by WardenGnaw** [comment]
> @gregg-miskelly Is it normal to require `dotnet restore` before building the solution? The building experience is a bit weird in VS since you have to do the restore first before building the solution.
> 
> There also seems to be an issue with `MakePIAPortable` and `WindowsDebugLauncher` not restoring when you run `dotnet restore src\MIDebugEngine.sln`

Looks like I need to do something like `<Target Name="RestoreSDKProjects"`

**Comment by gregg-miskelly** [comment]
> @gregg-miskelly Is it normal to require `dotnet restore` before building the solution? The building experience is a bit weird in VS since you have to do the restore first before building the solution.

The way it should work is that if you open the solution it should start restoring in the background. If you aren't seeing that, do you have "Tools->Options->NuGet Package Manager->Allow NuGet to download missing packages" and "Automatically check for missing packages during build in Visual Studio" checked?

It might be expected that if you build on the command line, and didn't do a `dotnet restore` first, it will fail. I thought they changed that, but maybe only for certain project types.

> There also seems to be an issue with `MakePIAPortable` and `WindowsDebugLauncher` not restoring when you run `dotnet restore src\MIDebugEngine.sln`

This is probably because they aren't configured to build by default. Can you right click on the solution or project and restore them?

File: src/SSHDebugTests/SSHDebugUnitTests.csproj [hunk 3]
```diff
@@ -65,6 +65,11 @@
       <Project>{15bcbef4-1c2b-412b-925b-34a049097e62}</Project>
       <Name>SSHDebugPS</Name>
     </ProjectReference>
+    <ProjectReference Include="..\DebugEngineHost.Stub\DebugEngineHost.Stub.csproj">
```
Comment (WardenGnaw) [inline src/SSHDebugTests/SSHDebugUnitTests.csproj:68]: I don't know if this is a bug but it is already declared false in:
https://github.com/microsoft/MIEngine/pull/1075/files#diff-34e5ea0482926d6a8c626f9912d5786f1c996e27b3016be4b1bdcee5c6ff8518R117-R132

This was needed to stop https://dev.azure.com/ms/MIEngine/_build/results?buildId=120543&view=logs&j=9d538be6-44f8-50c3-7930-4e7f88a0f6b2&t=86540a71-1fda-5070-7b33-ef0a45c48f8e&l=971 and make the unit tests work.

**Comment by Trass3r** [comment]
This paves the way for a .NET Core build of MIEngine, am I right?
Maybe even for Alpine support by getting rid of Mono?

**Comment by WardenGnaw** [comment]
> This paves the way for a .NET Core build of MIEngine, am I right?
> Maybe even for Alpine support by getting rid of Mono?

Yep. The idea is to first move to SDK style projects but still build with .NET Framework Leaving the actual binaries unchanged. Then there will be work needed to convert the Desktop.Release (XPLAT) portion to build and publish using .NET Core.
By doing so we can remove the Mono dependency and can publish OpenDebugAD7 on any Runtime .NET Core supports.

See https://docs.microsoft.com/en-us/dotnet/core/rid-catalog#using-rids

**Comment by WardenGnaw** [comment]
@gregg-miskelly When you have the time, can you review? The bigger changes are in https://github.com/microsoft/MIEngine/pull/1075/commits/21aa1994e46cf5896cbdeac45da6e5cb2075e047
which include creating a VSIX and auto create the pkgdef.

File: src/MIDebugPackage/MIDebugPackage.csproj [hunk 4]
```diff
@@ -1,183 +1,50 @@
...
-  <PropertyGroup Condition=" '$(VsSDKInstall)' != '' ">
-    <VsSDKIncludes Condition=" '$(VsSDKIncludes)' == '' ">$(VsSDKInstall)\VisualStudioIntegration\Common\inc</VsSDKIncludes>
-    <VsSDKToolsPath Condition=" '$(VsSDKToolsPath)' == '' ">$(VsSDKInstall)\VisualStudioIntegration\Tools\bin</VsSDKToolsPath>
+    <GeneratePkgDefFile>true</GeneratePkgDefFile>
+    <IncludeAssemblyInVSIXContainer>false</IncludeAssemblyInVSIXContainer>
+    <IncludeDebugSymbolsInVSIXContainer>true</IncludeDebugSymbolsInVSIXContainer>
+    <IncludeDebugSymbolsInLocalVSIXDeployment>true</IncludeDebugSymbolsInLocalVSIXDeployment>
   </PropertyGroup>
+
   <ItemGroup>
```
Comment (gregg-miskelly) [inline src/MIDebugPackage/MIDebugPackage.csproj:19]: >ItemGroup [](start = 3, length = 9)

Do you need: `<PackageReference Include="Microsoft.VisualStudioEng.MicroBuild.Core" Version="$(Microsoft_VisualStudioEng_MicroBuild_Core_Version)" GeneratePathProperty="true"/>`

Comment (gregg-miskelly) [reply src/MIDebugPackage/MIDebugPackage.csproj:19]: Nevermind. You have it

---
In reply to: [538863156](https://github.com/microsoft/MIEngine/pull/1075#discussion_r538863156) [](ancestors = 538863156)

File: src/MIDebugPackage/MIDebugPackage.csproj [hunk 5]
```diff
@@ -1,183 +1,50 @@
...
-    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <TargetFramework>net472</TargetFramework>
     <RootNamespace>Microsoft.MIDebugPackage</RootNamespace>
     <AssemblyName>Microsoft.MIDebugPackage</AssemblyName>
-    <TargetFrameworkVersion>v4.6.2</TargetFrameworkVersion>
-  </PropertyGroup>
-  <PropertyGroup>
-    <!-- Because the MIEngine ships in Visual Studio, it is recommended NOT to install the .vsix
-         Instead, The Visual Studio files should be directly updated.
-         The VSIX is built because it is useful for F5 from VS targeting the experimental instance. -->
```
Comment (gregg-miskelly) [inline src/MIDebugPackage/MIDebugPackage.csproj:81]: Put back this comment?

File: src/MIDebugPackage/MIDebugPackage.csproj [hunk 6]
```diff
@@ -1,183 +1,50 @@
...
-    <StartAction>Program</StartAction>
-    <StartProgram Condition="'$(DevEnvDir)' != ''">$(DevEnvDir)devenv.exe</StartProgram>
-    <StartArguments>/rootsuffix Exp</StartArguments>
-  </PropertyGroup>
-  <!-- This is to fix a build issue where VSSDK BuildTools are not able to be found. -->
-  <PropertyGroup Condition=" '$(VsSDKInstall)' != '' ">
-    <VsSDKIncludes Condition=" '$(VsSDKIncludes)' == '' ">$(VsSDKInstall)\VisualStudioIntegration\Common\inc</VsSDKIncludes>
-    <VsSDKToolsPath Condition=" '$(VsSDKToolsPath)' == '' ">$(VsSDKInstall)\VisualStudioIntegration\Tools\bin</VsSDKToolsPath>
+    <GeneratePkgDefFile>true</GeneratePkgDefFile>
+    <IncludeAssemblyInVSIXContainer>false</IncludeAssemblyInVSIXContainer>
```
Comment (gregg-miskelly) [inline src/MIDebugPackage/MIDebugPackage.csproj:14]: ><IncludeAssemblyInVSIXContainer>false</IncludeAssemblyInVSIXContainer> [](start = 4, length = 70)

Do you know why this is turned off?

**Comment by gregg-miskelly** [comment]
>     <IntermediateGeneratedOutputPath>$(IntermediateOutputPath)Generated\</IntermediateGeneratedOutputPath>

(This is still active)

---
In reply to: [721445297](https://github.com/microsoft/MIEngine/pull/1075#issuecomment-721445297) [](ancestors = 721445297)

---
Refers to: build/all_projects.settings.targets:23 in 21aa199. [](commit_id = 21aa1994e46cf5896cbdeac45da6e5cb2075e047, deletion_comment = False)

File: build/DropFiles.targets [hunk 7]
```diff
@@ -3,7 +3,6 @@
   <ItemGroup>
     <FilesToSign Include="@(DropSignedFile)">
       <Authenticode>Microsoft400</Authenticode>
-      <StrongName>StrongName</StrongName>
```
Comment (gregg-miskelly) [inline build/DropFiles.targets:6]: ><StrongName>StrongName</StrongName> [](start = 6, length = 35)

Is this set somewhere else?

**Comment by gregg-miskelly** [comment]
> @gregg-miskelly Is it normal to require `dotnet restore` before building the solution? The building experience is a bit weird in VS since you have to do the restore first before building the solution.
> 
> There also seems to be an issue with `MakePIAPortable` and `WindowsDebugLauncher` not restoring when you run `dotnet restore src\MIDebugEngine.sln`

Andrew, I missed your comment originally. Is this still an issue? Normally if you open a .NET SDK-style solution, the solution will automatically restore in the background after you open it. I *think* doing a build from the IDE will wait for it to finish.

File: build/DropFiles.targets (hunk 7 above)
Comment (WardenGnaw) [reply build/DropFiles.targets:6]: Good catch, I was doing this to test release builds.

**Comment by WardenGnaw** [comment]
> > @gregg-miskelly Is it normal to require `dotnet restore` before building the solution? The building experience is a bit weird in VS since you have to do the restore first before building the solution.
> > There also seems to be an issue with `MakePIAPortable` and `WindowsDebugLauncher` not restoring when you run `dotnet restore src\MIDebugEngine.sln`
> 
> Andrew, I missed your comment originally. Is this still an issue? Normally if you open a .NET SDK-style solution, the solution will automatically restore in the background after you open it. I _think_ doing a build from the IDE will wait for it to finish.

Resolved. I think VS does something weird and caches the csproj style in the `.vs` folder. So if you go from old style to sdk style without removing the `.vs` folder, it skips the restore step. 

Thanks for following up :)

**Comment by gregg-miskelly** [comment]
Otherwise LGTM

File: src/MIDebugPackage/MIDebugPackage.csproj (hunk 6 above)
Comment (WardenGnaw) [reply src/MIDebugPackage/MIDebugPackage.csproj:14]: Template from somewhere else, this seem to no copy MIDebugPackage.dll into the vsix.

I will enable this.

File: build/package_versions.settings.targets [hunk 8]
```diff
@@ -36,6 +36,7 @@
         <Microsoft_VisualStudio_Shell_Interop_8_0_Version>8.0.50728</Microsoft_VisualStudio_Shell_Interop_8_0_Version>
         <Microsoft_VisualStudio_Shell_Interop_9_0_Version>9.0.30730</Microsoft_VisualStudio_Shell_Interop_9_0_Version>
         <Microsoft_VisualStudio_Shell_Interop_Version>7.10.6072</Microsoft_VisualStudio_Shell_Interop_Version>
+        <Microsoft_VisualStudio_Shell_Interop_Version>7.10.6072</Microsoft_VisualStudio_Shell_Interop_Version>
```
Comment (gregg-miskelly) [inline build/package_versions.settings.targets:39]: This line is repeated

**Review by gregg-miskelly — APPROVED** [review]
Aside from that one nit, looks good to me.

File: build/DropFiles.targets (hunk 7 above)
Comment (WardenGnaw) [reply build/DropFiles.targets]: I think I need to Condition this  if SignType == 'real'.

File: build/package_versions.settings.targets [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        <Microsoft_VisualStudio_Threading_Version>15.3.23</Microsoft_VisualStudio_Threading_Version>
+        <Microsoft_VisualStudio_Utilities_Version>15.0.26606</Microsoft_VisualStudio_Utilities_Version>
+        <Microsoft_VisualStudio_Shell_Interop_15_0_DesignTime_Version>15.0.26932</Microsoft_VisualStudio_Shell_Interop_15_0_DesignTime_Version>
+        <Microsoft_VisualStudio_Workspace_Version>15.0.392</Microsoft_VisualStudio_Workspace_Version>
+        <Microsoft_VisualStudio_Workspace_VSIntegration_Version>15.0.392</Microsoft_VisualStudio_Workspace_VSIntegration_Version>
+        <Microsoft_VisualStudio_TextManager_Interop_Version>16.7.30328.74</Microsoft_VisualStudio_TextManager_Interop_Version>
+        <Microsoft_VSSDK_BuildTools_Version>16.7.3069</Microsoft_VSSDK_BuildTools_Version>
+        <envDTE_Version>8.0.2</envDTE_Version>
+        <System_Text_Json_Version>4.6.0</System_Text_Json_Version>
```
Comment (WardenGnaw) [inline build/package_versions.settings.targets:48]: Unused.

File: build/DropFiles.targets (hunk 7 above)
Comment (gregg-miskelly) [reply build/DropFiles.targets:6]: Is this the right condition, or should it really be lab configuration based? I thought you needed this for test signing support.

Comment (gregg-miskelly) [reply build/DropFiles.targets:6]: Oh, I see. Nevermind.

**Comment by WardenGnaw** [comment]
Validating the Releases:
| Status | Task |
| --- | :-- |
| ❔  | VS Release |
| ❔  |   VS Code Release |

**Pull request merged**